docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,2979 @@ p4grammar {
+- An underscore (that is, $\_$) is used when we do not need to use some specific information returned by some helper in judgments. 
+- The notation $\maybe \anyTyp$ denotes an optional type, that is, the type of $\anyTyp$ has been extended with bottom (maybe type in Haskell, or option type in OCaml). So if $\anyTyp$ denotes an integer, $\maybe \anyTyp$ could either be an integer number or bottom. 
+
+The following describes the metavariables used in this document:
+
+~ Center
+| **metavariable**   |   **type represented** |
+|:----------------:|:------------------|
+| $\bool$ | booleans |
+| $\str$ | strings |
+| $\int$ | arbitrary length integer |
+| $\width$ | width of a data structure |
+| $\size$ | size of a data structure |
+| $\bitWidth \bit \width$ | fixed length integer |
+| $\intWidth \int \width$ | signed integer |
+| $\name$ | strings representing any kinds of names |
+| $\typ$ | types |
+| $\dir$ | directions |
+| $\typVar$ | type variables |
+| $\wildcardParam$ | wild card type parameters |
+| $\field$ | field names |
+| $\env$ | environments |
+| $\emp$ | empty list of elements |
+| $\anyTyp$ | a value of any type |
+| $\constraint$, $\optConstraint$, $\nonoptConstraint$ | constraints |
+~
 
-1. It type checks P4 programs.
-2. It conducts type inference.
-3. It does a pass from the surface syntax to the first IR. 
 
+## Signatures
+``TODO``
+function signature. judgment signature. environment signature. 
+
+## contexts and environments
+
+# Syntax { #sec-syntax }
+
+In the following we present P4 syntax. Note that we omitted annotations and meta-information since they are not required for type checking.
+
+~ Begin P4Syntax
+**Data types used in P4 programs:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\baseTyp$ | $::=$ | $\boolTyp$ | $\qquad \text{booleans}$ |
+|   | $\vert$ | $\errTyp$ | $\qquad \text{errors}$ |
+|   | $\vert$ | $\integerTyp$ | $\qquad \text{infinite-precision integers}$ |
+|   | $\vert$ | $\intWidthTyp \exp$ | $\qquad \text{fixed-width signed integers}$ |
+|   | $\vert$ | $\bitWidthTyp \exp$ | $\qquad \text{fixed-width unsigned integers}$ |
+|   | $\vert$ | $\varBitTyp \exp$ | $\qquad \text{variable-width integers with a maximum width}$ |
+|   | $\vert$ | $\typVar$ | $\qquad \text{type variables}$ |
+|   | $\vert$ | $\spcTyp \baseTyp \baseTyps$ | $\qquad \text{specialized type}$ |
+|   | $\vert$ | $\stackTyp \baseTyp \exp$ | $\qquad \text{stacks}$ |
+|   | $\vert$ | $\tupleTyps \baseTyps$ | $\qquad \text{tuples}$ |
+|   | $\vert$ | $\stringTyp$ | $\qquad \text{strings}$ |
+|   | $\vert$ | $\voidTyp$ | $\qquad \text{void}$ |
+|   | $\vert$ | $\dontcareTyp$ | $\qquad \text{don't care}$ |
+
+**Types of P4 expressions:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\typ$ | $::=$ | $\boolTyp$ | $\qquad \text{booleans}$ |
```

Comment:
Pedantic, but where there are types in common it'd be nice for them to be listed in the same order as the grammar above.

Reply:
@pataei apply this in your last review of the syntax.